test/borda_clj/trace.clj

```diff
@@ -0,0 +1,43 @@
+(ns borda-clj.trace
+  (:require  [clojure.test :as t]
+             [borda-clj.client :as sut]
+             [clojure.core.async :as async :refer (<! <!! >! >!! alts! chan close! go go-loop onto-chan timeout)]))
+
+(defn run-trace [global-dimensions max-buffer-size trace]
+  (let [[report-ch measurements-ch kill-ch timeout-ch trace-ch] (repeatedly chan)
+        successful-send (fn [m] (>!! trace-ch {:send m}))
+        failed-send (fn [& args] (throw (RuntimeException. "boom")))
+        final-send (fn [m] (>!! trace-ch {:final-send m}))
+        send (fn [measurements] ((or (:trace-metadata (meta measurements))
+                                    final-send)
+                                measurements))
+        on-send-error (fn [m e]
+                        ;;(clojure.pprint/pprint e)
+                        (>!! trace-ch {:error {:measurements m}}))
+        flush-ch (sut/go-flush-measurements report-ch measurements-ch kill-ch global-dimensions ::bogus-interval send on-send-error (constantly timeout-ch))]
+    (sut/go-reduce-reports report-ch measurements-ch kill-ch max-buffer-size)
+    (go
+      (<! flush-ch)
+      (close! trace-ch))
+    (go
+      (doseq [op trace]
+        (case op
+          :successful-tick (>! timeout-ch successful-send)
+          :failed-tick (>! timeout-ch failed-send)
+          :delay (<! (timeout 100))
+          (>! report-ch op)))
+      (<! (timeout 100))
+      (close! report-ch))
+    (<!! (async/into [] trace-ch))))
+
+(comment
+
+  (clojure.pprint/pprint
+   (run-trace {:g "global"}
+              2
+              [[{:d "d"} {:v 1}]
+               [{:e "e"} {:v 1}]
+               [{:f "f"} {:v 1}]
+               :failed-tick]))
```

Comment:
There's an implicit stop and flush at the end.  The result of it can be ignored by filtering out `:final-send` entries

This particular run prints

```
borda buffer full, discarding measurement {:f f} {:_submits 1, :v 1}
borda buffer full, discarding measurement {:g global, :op _submit} {:_submits 1, :error_count 1, :success_count 2}
[{:error
  {:measurements
   {{:g "global", :d "d"} {:v 1},
    {:g "global", :e "e"} {:v 1},
    {:g "global", :op "_submit"} {:error_count 1, :success_count 2}}}}
 {:final-send
  {{:g "global", :d "d"} {:v 1},
   {:g "global", :e "e"} {:v 1},
   {:g "global", :op "_submit"} {:error_count 1, :success_count 2}}}]
```

Reply:
> We seem to have a bug whereby {:op "_submit"} measurements may be lost in resubmitting. This will be less probable with higher max-buffer-size, but it's still a possibility. Maybe we should whitelist this key?

Hmm, I thought I had already whitelisted it [here](https://github.com/getlantern/borda-clj/blob/master/src/borda_clj/client.clj#L22) and tested that it's working with the conditions [here](https://github.com/getlantern/borda-clj/blob/master/test/borda_clj/client_test.clj#L28)?

Reply:
@oxtoacart looks good.  This comment may have happened before that or I may have missed it -- sorry for the noise.